source/fe/mapping_manifold.cc

```diff
@@ -0,0 +1,1308 @@
+                      const typename dealii::Triangulation<1,spacedim>::cell_iterator &cell,
+                      const unsigned int &)
+    {
+      return cell->get_manifold();
+    }
+
+    /**
+     * Some specialization for face Manifolds. The mapping argument is
+     * only used to help the compiler infer dim and spacedim.
+     */
+    template<int dim, int spacedim>
+    const dealii::Manifold<dim,spacedim> &
+    get_face_manifold(const MappingManifold<dim,spacedim> &,
+                      const typename dealii::Triangulation<dim,spacedim>::cell_iterator &cell,
+                      const unsigned int face_no)
+    {
+      return cell->face(face_no)->get_manifold();
+    }
+
+    /**
+     * Compute the locations of quadrature points on the object described by
+     * the first argument (and the cell for which the mapping support points
+     * have already been set), but only if the update_flags of the @p data
+     * argument indicate so.
+     */
+    template <int dim, int spacedim>
+    void
+    maybe_compute_q_points (const typename QProjector<dim>::DataSetDescriptor                 data_set,
+                            const typename dealii::MappingManifold<dim,spacedim>::InternalData      &data,
+                            std::vector<Point<spacedim> >                                     &quadrature_points)
+    {
+      const UpdateFlags update_flags = data.update_each;
+
+      AssertDimension(data.vertices.size(), GeometryInfo<dim>::vertices_per_cell);
+
+      if (update_flags & update_quadrature_points)
+        {
+          for (unsigned int point=0; point<quadrature_points.size(); ++point)
+            {
+              quadrature_points[point] = data.manifold->
+                                         get_new_point(Quadrature<spacedim>(data.vertices,
+                                                       data.cell_manifold_quadrature_weights[point+data_set]));
+            }
+        }
+    }
+
+
+    /**
+     * Update the co- and contravariant matrices as well as their determinant, for the cell
+     * described stored in the data object, but only if the update_flags of the @p data
+     * argument indicate so.
+     */
+    template <int dim, int spacedim>
+    void
+    maybe_update_Jacobians (const typename dealii::QProjector<dim>::DataSetDescriptor          data_set,
+                            const typename dealii::MappingManifold<dim,spacedim>::InternalData      &data)
+    {
+      const UpdateFlags update_flags = data.update_each;
+
+      if (update_flags & update_contravariant_transformation)
+        {
+          const unsigned int n_q_points = data.contravariant.size();
+
+          std::fill(data.contravariant.begin(), data.contravariant.end(),
+                    DerivativeForm<1,dim,spacedim>());
+
+          // Cache of weights used to compute points on the reference cell
+          std::vector<double> weights(GeometryInfo<dim>::vertices_per_cell);
```

Comment:
We've tried to make these 'cache' variables mutable members of the internal data object to avoid the repeated memory allocation. Could you move it there?